realm/realm-library/src/main/java/io/realm/HandlerController.java

```diff
@@ -146,6 +151,23 @@ public boolean handleMessage(Message message) {
         return true;
     }
 
+    /**
+     * Properly handles when an async transaction completes. This will be treated as a REALM_CHANGED event when
+     * determining which queries to re-run and when to notify listeners.
+     * <p>
+     * NOTE: This is needed as it is not possible to combine a `Message.what` value and a callback runnable. So instead
+     * of posting two messages, we post a runnable that runs this method. This means it is possible to interpret
+     * `REALM_CHANGED + Runnable` as one atomic message.
+     *
+     * @param onSuccess onSuccess callback to run for the async transaction that completed.
+     */
+    public void handleAsyncTransactionCompleted(Runnable onSuccess) {
+        if (onSuccess != null) {
+            pendingOnSuccessAsyncTransactionCallbacks.add(onSuccess);
+        }
+        realmChanged(false);
```

Comment:
More thoughts about this, thread safe container is not enough. We need to lock before `advance_read` happen, or say, we should have a copy of the callback list before `advance_read` happen. Any callbacks posted later after this `advance_read` should be called after next `advance_read`.